phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -247,9 +255,12 @@ private void repairIndexRows(byte[] indexRowKey, long ts, List<Cell> row) throws
                 buildIndexScan.setAttribute(BaseScannerRegionObserver.SKIP_REGION_BOUNDARY_CHECK, Bytes.toBytes(true));
             }
             // Rebuild the index rows from the corresponding the rows in the the data table
+            // Get the data row key from the index row key
             byte[] dataRowKey = indexMaintainer.buildDataRowKey(new ImmutableBytesWritable(indexRowKey), viewConstants);
+            // Rebuild the index rows from the data rows starting with the data row key
             buildIndexScan.withStartRow(dataRowKey, true);
-            buildIndexScan.setTimeRange(ts, maxTimestamp);
+            buildIndexScan.setTimeRange(0, maxTimestamp);
```

Comment:
I'm a bit confused by the startTime here. I'd have expected it to be either {0, Long.MAX_VALUE} to look at all rows, {0, EnvironmentEdgeManager.currentTimeMillis()} to look up to the current server timestamp, or {Math.max(0, minTimestamp), Math.min(maxTimestamp, Long.MAX_VALUE}, to look at the timestamps visible to the user Scan, where minTimestamp and maxTimestamp come from, but not a mixture of 0 and maxTimestamp. 

If we do honor the minTimestamp, we need to make sure not to delete from the index based on not finding a data row that might be hidden by the minTimestamp.

Reply:
The min and max timestamp are from the original scan on the index (set by the client). I did not want to honor the minTimestamp to cover one case. Assume that there are bunch of overwrites on the same row that made only to the index table within min and max timestamps. If we build index table using [minTimestamp, maxTimestamp], then the data table row will not be visible to buildIndexScan. Also assume that the compaction process has removed the good version of this index row (i.e., verified version). If so, we may end up deleting all the versions of the index row (if we start from minTimestamp).  I think Phoenix client will always set minTimestamp to zero or table create timestamp but I want to be on the safe side.

Reply:
Right, I get that there are corner cases where honoring minTimestamp would cause problems, so you're choosing 0. Why are you honoring maxTimestamp during the rebuild? Why should the SCN a client has chosen affect the rebuild semantics? (As opposed to the client's query results, which should of course be affected by SCN)

Reply:
There is no reason to that. If we do that we would rebuild rows unnecessarily. Assume that there are N overwrite attempts on a particular row and all of them failed at phase 2 except the last overwrite. Also assume that just before the successful one, a client scans the index table. This scan will see the index row is unverified and will rebuilt the older version and will get the older successful version (not the last successful version), which is correct behavior. If we do not honor maxTimestamp then we would build both the old and new version but return to the client old version. In other words, we unnecessarily would build the newer versions even though they may not need to be rebuilt.